lib/trivial.nix

```diff
@@ -334,7 +334,10 @@ rec {
      has the same return type and semantics as builtins.functionArgs.
      setFunctionArgs : (a → b) → Map String Bool.
   */
-  functionArgs = f: f.__functionArgs or (builtins.functionArgs f);
+  functionArgs = f:
+    if f ? __functor
+    then functionArgs (f.__functor f)
+    else (f.__functionArgs or builtins.functionArgs f);
```

Comment:
This doesn't work with `setFunctionArgs`:
```
nix-repl> functionArgs (setFunctionArgs (args: args.x) { x = false; })  
{ }
```

Also breaks completely for
```
nix-repl> functionArgs { __functionArgs = null; }
error: attempt to call something which is not a function but null, at /home/infinisil/src/nixpkgs/lib/trivial.nix:340:11
```

Because `f.__functionArgs or builtins.functionArgs f` == `(f.__functionArgs or builtins.functionArgs) f`.

Reply:
This should be implemented with
```nix
{
  functionArgs = f:
    if f ? __functor
    then f.__functionArgs or (lib.functionArgs f.__functor)
    else builtins.functionArgs f;
}
```
instead

Reply:
```nix

Reply:
As pointed out to me by @blaggacao on IRC, there's a problem with this:
```
nix-repl> functionArgs { __functor = self: { x }: x; }             
{ }
```

And we need to use `f.__functor f` instead

Reply:
[force pushed](https://github.com/NixOS/nixpkgs/compare/c024784605300f4975ed8825dfd3d610c4e694dd..cf8e219b7e3c8933d6301175f2611990c5281ae9)